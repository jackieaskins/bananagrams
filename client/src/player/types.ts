export type Player = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
};
