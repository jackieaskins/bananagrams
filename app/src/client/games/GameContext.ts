import { createContext, useContext } from "react";
import { GameState } from "./types";
import { Game } from "@/types/game";

export function getEmptyGameInfo(gameId: string): Game {
  return {
    gameId,
    gameName: "",
    isInProgress: false,
    bunch: [],
    players: [],
    previousSnapshot: null,
  };
}

export const GameContext = createContext<GameState>({} as GameState);

export function useGame(): GameState {
  return useContext(GameContext);
}
