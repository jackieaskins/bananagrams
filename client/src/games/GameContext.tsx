import React, { createContext, useContext, useEffect, useState } from 'react';
import { useHistory, useLocation, useParams } from 'react-router-dom';

import { useSocket } from '../SocketContext';
import { SetState } from '../state/types';

type GameParams = {
  gameId: string;
};

export type GameLocationState = {
  isInGame?: boolean;
  gameInfo?: GameInfo;
};

export type Player = {
  userId: string;
  username: string;
  isOwner: boolean;
  isReady: boolean;
};

export type GameInfo = {
  gameId: string;
  gameName: string;
  isInProgress: boolean;
  players: Player[];
};

type GameState = {
  gameId: string;
  gameName: string;
  isInGame: boolean;
  isInProgress: boolean;
  setIsInProgress: SetState<boolean>;
  setPlayers: SetState<Player[]>;
  setPlayerReady: (userId: string, isReady: boolean) => void;
  players: Player[];
};

const GameContext = createContext<GameState>({
  gameId: '',
  gameName: '',
  isInGame: false,
  isInProgress: false,
  setIsInProgress: () => undefined,
  setPlayerReady: () => undefined,
  setPlayers: () => undefined,
  players: [],
});

export const GameProvider: React.FC<{}> = ({ children }) => {
  const { socket } = useSocket();
  const { replace } = useHistory();
  const { pathname, state } = useLocation<GameLocationState>();
  const { gameId } = useParams<GameParams>();

  const [gameName] = useState(state?.gameInfo?.gameName ?? '');
  const [isInProgress, setIsInProgress] = useState(
    state?.gameInfo?.isInProgress ?? false
  );
  const [isInGame] = useState(state?.isInGame ?? false);
  const [players, setPlayers] = useState(state?.gameInfo?.players ?? []);

  const setPlayerReady = (userId: string, isReady: boolean): void => {
    setPlayers((prevPlayers) => {
      const index = prevPlayers.findIndex((player) => player.userId === userId);
      return [
        ...prevPlayers.slice(0, index),
        { ...prevPlayers[index], isReady },
        ...prevPlayers.slice(index + 1),
      ];
    });
  };

  useEffect(() => {
    if (isInGame) {
      replace(pathname);

      return (): void => {
        socket.emit('leaveGame', { gameId }, () => undefined);
      };
    }

    return;
  }, []);

  useEffect(() => {
    socket.on('playerJoined', (player: Player) => {
      setPlayers((prevPlayers) => [...prevPlayers, player]);
    });

    socket.on('playerLeft', ({ userId }: Player) =>
      setPlayers((prevPlayers) =>
        prevPlayers.filter((player) => player.userId !== userId)
      )
    );

    socket.on('playerReady', ({ userId }: Player) => {
      console.log(`${userId} is ready`);
      setPlayerReady(userId, true);
    });

    return (): void => {
      socket.off('playerJoined');
      socket.off('playerLeft');
      socket.off('playerReady');
    };
  }, []);

  return (
    <GameContext.Provider
      value={{
        gameId,
        gameName,
        isInGame,
        isInProgress,
        players,
        setIsInProgress,
        setPlayerReady,
        setPlayers,
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export const useGame = (): GameState => useContext(GameContext);
