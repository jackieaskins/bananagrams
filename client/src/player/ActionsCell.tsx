import { DeleteOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import { useCurrentPlayer } from '../game/stateHooks';
import { kickPlayer } from '../socket';
import { Player } from './types';

type ActionsCellProps = {
  player: Player;
};

const ActionsCell = ({
  player: { userId },
}: ActionsCellProps): JSX.Element | null => {
  const currentPlayer = useCurrentPlayer();

  if (!currentPlayer) {
    return null;
  }

  const { isAdmin, userId: currentPlayerId } = currentPlayer;

  if (isAdmin && userId !== currentPlayerId) {
    return (
      <Button
        icon={<DeleteOutlined />}
        onClick={() => {
          kickPlayer({ userId });
        }}
        type="text"
      />
    );
  }

  return null;
};

export default ActionsCell;
