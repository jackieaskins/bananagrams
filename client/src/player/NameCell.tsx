import { CrownOutlined, KeyOutlined } from '@ant-design/icons';
import { Space } from 'antd';

import { Player } from '../players/types';

type NameCellProps = {
  player: Player;
};

const NameCell = ({
  player: { isAdmin, isTopBanana, username },
}: NameCellProps): JSX.Element => (
  <Space>
    <span>{username}</span>
    {isAdmin && (
      <span>
        <KeyOutlined />
      </span>
    )}
    {isTopBanana && (
      <span>
        <CrownOutlined />
      </span>
    )}
  </Space>
);

export default NameCell;
