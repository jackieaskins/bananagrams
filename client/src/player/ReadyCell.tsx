import { CheckOutlined, CloseOutlined } from '@ant-design/icons';
import { Checkbox } from 'antd';

import { getUserId, setReady } from '../socket';
import { Player } from './types';

type ReadyCellProps = {
  player: Player;
};

const ReadyCell = ({
  player: { userId, isReady },
}: ReadyCellProps): JSX.Element => {
  const isCurrentUser = userId === getUserId();

  if (isCurrentUser) {
    return (
      <Checkbox
        checked={isReady}
        onChange={({ target: { checked } }) => {
          setReady({ isReady: checked });
        }}
      />
    );
  }

  if (isReady) {
    return <CheckOutlined css={{ color: 'green' }} />;
  }

  return <CloseOutlined css={{ color: 'red' }} />;
};
export default ReadyCell;
