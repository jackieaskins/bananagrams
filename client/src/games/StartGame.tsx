import { Box, Grid, Typography } from "@mui/material";
import OpponentBoardPreview from "../boards/OpponentBoardPreview";
import CopyToClipboard from "../buttons/CopyToClipboard";
import PlayerList from "../players/PlayerList";
import { useGame } from "./GameContext";

const StartGame: React.FC = () => {
  const {
    gameInfo: { gameName, previousSnapshot },
  } = useGame();
  const joinUrl = `${window.location.href}/join`;

  return (
    <Box>
      <Typography variant="h3" align="center" gutterBottom>
        {gameName}
      </Typography>
      <Typography variant="body1" align="center" color="textSecondary">
        Invite others to game:
      </Typography>
      <Typography
        variant="body2"
        align="center"
        color="textSecondary"
        gutterBottom
      >
        {joinUrl} <CopyToClipboard copyText={joinUrl} />
      </Typography>

      <Grid container justifyContent="center" spacing={3}>
        <Grid item md={5}>
          <PlayerList />
        </Grid>
        {previousSnapshot && (
          <Grid item>
            <Box display="flex" flexDirection="column">
              <Typography variant="body1" align="center" gutterBottom>
                Here are the boards from that round:
              </Typography>

              <OpponentBoardPreview
                initialPlayerIndex={previousSnapshot.findIndex(
                  (player) => player.isTopBanana,
                )}
                players={previousSnapshot}
                includeCurrentPlayer
              />
            </Box>
          </Grid>
        )}
      </Grid>
    </Box>
  );
};

export default StartGame;
