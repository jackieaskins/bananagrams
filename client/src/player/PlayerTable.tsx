import { Table, Typography } from 'antd';
import { ColumnType } from 'antd/lib/table';

import { useCurrentPlayer, useGamePlayers } from '../game/stateHooks';
import ActionsCell from './ActionsCell';
import NameCell from './NameCell';
import ReadyCell from './ReadyCell';
import { Player } from './types';

const COLUMN_ALIGN = 'center';
const MAX_PLAYERS = 8;

const { Text } = Typography;

const actionsColumn: ColumnType<Player> = {
  align: COLUMN_ALIGN,
  key: 'actions',
  render: function ActionsRender(_: any, player: Player) {
    return <ActionsCell player={player} />;
  },
};

const PlayerTable = (): JSX.Element => {
  const players = useGamePlayers();
  const isAdmin = !!useCurrentPlayer()?.isAdmin;

  return (
    <Table
      bordered
      columns={[
        {
          title: 'Players',
          children: [
            {
              align: COLUMN_ALIGN,
              key: 'ready',
              title: 'Ready?',
              render: function ReadyRender(_, player) {
                return <ReadyCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              key: 'player',
              title: 'Player',
              render: function PlayerRender(_, player) {
                return <NameCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              dataIndex: 'gamesWon',
              title: 'Games won',
            },
            ...(isAdmin && players.length > 1 ? [actionsColumn] : []),
          ],
        },
      ]}
      dataSource={players}
      footer={() => (
        <div css={{ fontSize: '12px', textAlign: 'center' }}>
          <Text type="secondary">
            The game will start when all players are ready
            <br />
            Current player count: {players.length} / {MAX_PLAYERS}
          </Text>
        </div>
      )}
      pagination={false}
      rowKey="userId"
      size="middle"
    />
  );
};

export default PlayerTable;
