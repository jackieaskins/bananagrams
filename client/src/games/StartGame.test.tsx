import { Button, useClipboard } from "@chakra-ui/react";
import { shallow } from "enzyme";
import { playerFixture } from "../fixtures/player";
import { useGame } from "./GameContext";
import StartGame from "./StartGame";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useEffect: jest.fn((fn) => fn()),
}));

jest.mock("./GameContext", () => ({
  useGame: jest.fn(),
}));

const mockToast = jest.fn();
const mockOnCopy = jest.fn();
jest.mock("@chakra-ui/react", () => ({
  ...jest.requireActual("@chakra-ui/react"),
  useClipboard: jest.fn(),
  useToast: jest.fn(() => mockToast),
}));

describe("<StartGame />", () => {
  const mockUseGame = (previousSnapshot = null) => {
    useGame.mockReturnValue({
      gameInfo: {
        gameName: "gameName",
        previousSnapshot,
      },
    });
  };

  const renderComponent = () => shallow(<StartGame />);

  beforeEach(() => {
    mockUseGame();
    useClipboard.mockReturnValue({ hasCopied: false, onCopy: mockOnCopy });
  });

  test("copies the invite link when the copy button is clicked", () => {
    renderComponent().find(Button).props().onClick();

    expect(mockOnCopy).toHaveBeenCalledWith();
  });

  test("shows a toast message when the invite link is copied", () => {
    useClipboard.mockReturnValue({ hasCopied: true, onCopy: mockOnCopy });

    const component = renderComponent();
    component.update();

    expect(mockToast).toHaveBeenCalledWith({
      description: "Successfully copied invite link to clipboard",
    });
  });

  test("does not show a toast message when invite link has not been copied", () => {
    useClipboard.mockReturnValue({ hasCopied: true, onCopy: mockOnCopy });

    const component = renderComponent();
    component.update();

    expect(mockToast).not.toHaveBeenCalledWith();
  });

  test("does not render board when no snapshot", () => {
    mockUseGame();

    expect(renderComponent()).toMatchSnapshot();
  });

  test("renders game boards when snapshot is present", () => {
    mockUseGame([playerFixture({ userId: "123", isTopBanana: true })]);

    expect(renderComponent()).toMatchSnapshot();
  });
});
