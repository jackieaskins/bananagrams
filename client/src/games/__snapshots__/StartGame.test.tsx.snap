// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StartGame /> does not render board when no snapshot 1`] = `
<Styled(MuiBox)>
  <WithStyles(ForwardRef(Typography))
    align="center"
    gutterBottom={true}
    variant="h3"
  >
    gameName
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    color="textSecondary"
    variant="body1"
  >
    Invite others to game:
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    color="textSecondary"
    gutterBottom={true}
    variant="body2"
  >
    http://localhost//join
     
    <CopyToClipboard
      copyText="http://localhost//join"
    />
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={5}
    >
      <PlayerList />
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Styled(MuiBox)>
`;

exports[`<StartGame /> renders winning board when snapshot is present 1`] = `
<Styled(MuiBox)>
  <WithStyles(ForwardRef(Typography))
    align="center"
    gutterBottom={true}
    variant="h3"
  >
    gameName
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    color="textSecondary"
    variant="body1"
  >
    Invite others to game:
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    color="textSecondary"
    gutterBottom={true}
    variant="body2"
  >
    http://localhost//join
     
    <CopyToClipboard
      copyText="http://localhost//join"
    />
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    justify="center"
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={5}
    >
      <PlayerList />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
    >
      <Styled(MuiBox)
        display="flex"
        flexDirection="column"
      >
        <WithStyles(ForwardRef(Typography))
          align="center"
          gutterBottom={true}
          variant="body1"
        >
          Here is 
          username's
           winning board:
        </WithStyles(ForwardRef(Typography))>
        <PreviewBoard
          board={
            Array [
              Array [
                null,
              ],
            ]
          }
          tileSize={20}
        />
      </Styled(MuiBox)>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Styled(MuiBox)>
`;
